src/sdf/interval.h

```diff
@@ -0,0 +1,207 @@
+    else if (upper <= 0.0 && other.upper <= 0.0)
+      return {a2b2, a1b1};
+
+    Domain a1b2 = lower * other.upper;
+    Domain a2b1 = upper * other.lower;
+    return {std::min(std::min(a1b1, a1b2), std::min(a2b1, a2b2)),
+            std::max(std::max(a1b1, a1b2), std::max(a2b1, a2b2))};
+  }
+
+  Interval operator*(double d) const {
+    if (d > 0) return {lower * d, upper * d};
+    return {upper * d, lower * d};
+  }
+
+  Interval operator/(const Interval &other) const {
+    if (other.is_const()) return *this / other.lower;
+    constexpr Domain zero = static_cast<Domain>(0);
+    constexpr Domain infty = std::numeric_limits<Domain>::infinity();
+    Interval reci;
+    if (other.lower >= zero || other.upper <= zero) {
+      reci.lower = other.upper == zero ? -infty : (1 / other.upper);
+      reci.upper = other.lower == zero ? infty : (1 / other.lower);
+    } else {
+      reci.lower = -infty;
+      reci.upper = infty;
+    }
+    return *this * reci;
+  }
+
+  Interval operator/(double d) const {
+    if (d > 0) return {lower / d, upper / d};
+    return {upper / d, lower / d};
+  }
+
+  constexpr bool is_const() const { return lower == upper; }
+
+  constexpr Interval operator==(const Interval &other) const {
+    if (is_const() && other.is_const() && lower == other.lower)
+      return constant(1);  // must be equal
+    if (lower > other.upper || upper < other.lower)
+      return constant(0);  // disjoint, cannot possibly be equal
+    return {0, 1};
+  }
+
+  constexpr bool operator==(double d) const { return is_const() && lower == d; }
+
+  constexpr Interval operator>(const Interval &other) const {
+    if (lower > other.upper) return constant(1);
+    if (upper < other.lower) return constant(0);
+    return {0, 1};
+  }
+
+  constexpr Interval operator<(const Interval &other) const {
+    if (upper < other.lower) return constant(1);
+    if (lower > other.upper) return constant(0);
+    return {0, 1};
+  }
+
+  constexpr Interval min(const Interval &other) const {
+    return {std::min(lower, other.lower), std::min(upper, other.upper)};
+  }
+
+  constexpr Interval max(const Interval &other) const {
+    return {std::max(lower, other.lower), std::max(upper, other.upper)};
+  }
+
+  constexpr Interval merge(const Interval &other) const {
+    return {std::min(lower, other.lower), std::max(upper, other.upper)};
+  }
+
+  template <typename F>
+  constexpr Interval monotone_map(F f) const {
+    if (is_const()) return constant(f(lower));
+    return {f(lower), f(upper)};
+  }
+
+  template <typename F>
+  constexpr Interval antimonotone_map(F f) const {
+    if (is_const()) return constant(f(lower));
+    return {f(upper), f(lower)};
+  }
+
+  constexpr Interval abs() const {
+    if (lower >= 0) return *this;
+    if (upper <= 0) return {-upper, -lower};
+    return {0.0, std::max(-lower, upper)};
+  }
+
+  constexpr Interval mod(double m) const {
+    // FIXME: cannot deal with negative m right now...
```

Comment:
This is not the typical `std::fmod` in c++, this returns something positive. This is nicer to interval arithmetic, and IMO more natural. For negative `m` however, I am not sure what I should do with that, maybe give something negative?